subsys/net/lib/download_client/src/download_client.c

```diff
@@ -483,6 +496,21 @@ void download_thread(void *client, void *a, void *b)
 wait_for_download:
 	k_sem_take(&dl->wait_for_download, K_FOREVER);
 
+	if (dl->fd == -1) {
+		rc = client_connect(dl);
+
+		if (rc == 0) {
+			if (IS_ENABLED(CONFIG_COAP) &&
+			    (dl->proto == IPPROTO_UDP || dl->proto == IPPROTO_DTLS_1_2)) {
+				coap_block_init(client, dl->progress);
+			}
+			rc = request_send(dl);
+		}
+		if (rc) {
+			error_evt_send(dl, rc);
```

Comment:
Looks like `rc` can be both a positive error code from the socket operations and a negative error code we generate ourselves. `error_evt_send()` will assert on negative error codes.
It would be useful to know why we failed to determine if we could try again.

Reply:
Where do we get positive values?
`http_request_send()`, `coap_request_send()`  and `socket_send()` all seem to `return 0;` when things work. Positive values from `send()` are eaten in the `socket_send()`.

Reply:
I don't remember, but it was probably from connect. I think I tried a non existing URI, probably by changing the port number. After the timeout I got the Assert.

Reply:
`connect()` can only return 0 or -1.

I cannot spot any errors, I tried:
* Correct HTTP address
* Correct HTTPS address
* Incorrect HTTP and HTTPS addres
* Non existent domain

```
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Waiting for network.. OK
Downloading http://speedtest.ftp.otenet.gr/files/test100k.db
[ 100% ] |==================================================| (102400/102400 bytes)
Download completed in 9573 ms @ 10696 bytes per sec, total 102400 bytes
Bye
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Provisioning certificate
Waiting for network.. OK
Downloading https://nrfconnectsdk.s3.eu-central-1.amazonaws.com/sample-img-100kb.png
[ 100% ] |==================================================| (102923/102923 bytes)
Download completed in 41672 ms @ 2469 bytes per sec, total 102923 bytes
Bye
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Waiting for network.. OK
Downloading http://example.com:1234/nonexistent
Error 116 during download
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Certificate match
Waiting for network.. OK
Downloading https://example.com:1234/nonexistent
Error 116 during download
*** Booting Zephyr OS build v3.2.99-ncs2-33-ge414f0842616 ***
Download client sample started
Certificate match
Waiting for network.. OK
Downloading http://nonexistend.domain/nonexistent
Error 118 during download

```

All usecases seem to return sensible errors.

Reply:
Oh.. now I see.
I overlooked this, as I was claiming that none of the mentioned functions return positive erros, I did not notice that Lucas already told me that `error_evt_send()` expects positive. In my mind I was thinking opposite.

Now it is fixed. I convert the error code to positive for the event handler, which in turn convers it to negative..